File: android_app/app/src/main/java/com/oceansentinels/app/data/local/preferences/PreferencesManager.kt
package com.oceansentinels.app.data.local.preferences

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import androidx.datastore.preferences.preferencesDataStore
import com.google.gson.Gson
import com.oceansentinels.app.domain.model.User
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "ocean_sentinels_preferences")

/**
 * DataStore preferences manager for storing user session and app settings
 */
@Singleton
class PreferencesManager @Inject constructor(
    @ApplicationContext private val context: Context,
    private val gson: Gson
) {
    private val dataStore = context.dataStore
    
    companion object {
        private val KEY_AUTH_TOKEN = stringPreferencesKey("auth_token")
        private val KEY_USER_DATA = stringPreferencesKey("user_data")
        private val KEY_IS_LOGGED_IN = booleanPreferencesKey("is_logged_in")
        private val KEY_REMEMBER_ME = booleanPreferencesKey("remember_me")
        private val KEY_DARK_MODE = booleanPreferencesKey("dark_mode")
        private val KEY_NOTIFICATIONS_ENABLED = booleanPreferencesKey("notifications_enabled")
        private val KEY_FCM_TOKEN = stringPreferencesKey("fcm_token")
        private val KEY_FIRST_LAUNCH = booleanPreferencesKey("first_launch")
        private val KEY_DEMO_POPUP_SHOWN = booleanPreferencesKey("demo_popup_shown")
        private val KEY_OFFLINE_MODE = booleanPreferencesKey("offline_mode")
    }
    
    // ============= Auth Token =============
    
    val authToken: Flow<String?> = dataStore.data.map { preferences ->
        preferences[KEY_AUTH_TOKEN]
    }
    
    suspend fun saveAuthToken(token: String) {
        dataStore.edit { preferences ->
            preferences[KEY_AUTH_TOKEN] = token
        }
    }
    
    suspend fun clearAuthToken() {
        dataStore.edit { preferences ->
            preferences.remove(KEY_AUTH_TOKEN)
        }
    }
    
    // ============= User Data =============
    
    val userData: Flow<User?> = dataStore.data.map { preferences ->
        preferences[KEY_USER_DATA]?.let { json ->
            try {
                gson.fromJson(json, User::class.java)
            } catch (e: Exception) {
                null
            }
        }
    }
    
    suspend fun saveUserData(user: User) {
        dataStore.edit { preferences ->
            preferences[KEY_USER_DATA] = gson.toJson(user)
        }
    }
    
    suspend fun clearUserData() {
        dataStore.edit { preferences ->
            preferences.remove(KEY_USER_DATA)
        }
    }
    
    // ============= Login State =============
    
    val isLoggedIn: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_IS_LOGGED_IN] ?: false
    }
    
    suspend fun setLoggedIn(loggedIn: Boolean) {
        dataStore.edit { preferences ->
            preferences[KEY_IS_LOGGED_IN] = loggedIn
        }
    }
    
    // ============= Remember Me =============
    
    val rememberMe: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_REMEMBER_ME] ?: false
    }
    
    suspend fun setRememberMe(remember: Boolean) {
        dataStore.edit { preferences ->
            preferences[KEY_REMEMBER_ME] = remember
        }
    }
    
    // ============= Dark Mode =============
    
    val isDarkMode: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_DARK_MODE] ?: false
    }
    
    suspend fun setDarkMode(enabled: Boolean) {
        dataStore.edit { preferences ->
            preferences[KEY_DARK_MODE] = enabled
        }
    }
    
    // ============= Notifications =============
    
    val notificationsEnabled: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_NOTIFICATIONS_ENABLED] ?: true
    }
    
    suspend fun setNotificationsEnabled(enabled: Boolean) {
        dataStore.edit { preferences ->
            preferences[KEY_NOTIFICATIONS_ENABLED] = enabled
        }
    }
    
    // ============= FCM Token =============
    
    val fcmToken: Flow<String?> = dataStore.data.map { preferences ->
        preferences[KEY_FCM_TOKEN]
    }
    
    suspend fun saveFcmToken(token: String) {
        dataStore.edit { preferences ->
            preferences[KEY_FCM_TOKEN] = token
        }
    }
    
    // ============= First Launch =============
    
    val isFirstLaunch: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_FIRST_LAUNCH] ?: true
    }
    
    suspend fun setFirstLaunchComplete() {
        dataStore.edit { preferences ->
            preferences[KEY_FIRST_LAUNCH] = false
        }
    }
    
    // ============= Demo Popup =============
    
    val hasDemoPopupShown: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_DEMO_POPUP_SHOWN] ?: false
    }
    
    suspend fun setDemoPopupShown() {
        dataStore.edit { preferences ->
            preferences[KEY_DEMO_POPUP_SHOWN] = true
        }
    }
    
    // ============= Offline Mode =============
    
    val isOfflineMode: Flow<Boolean> = dataStore.data.map { preferences ->
        preferences[KEY_OFFLINE_MODE] ?: false
    }
    
    suspend fun setOfflineMode(enabled: Boolean) {
        dataStore.edit { preferences ->
            preferences[KEY_OFFLINE_MODE] = enabled
        }
    }
    
    // ============= Clear All =============
    
    suspend fun clearAll() {
        dataStore.edit { preferences ->
            preferences.clear()
        }
    }
    
    suspend fun clearSession() {
        dataStore.edit { preferences ->
            preferences.remove(KEY_AUTH_TOKEN)
            preferences.remove(KEY_USER_DATA)
            preferences[KEY_IS_LOGGED_IN] = false
        }
    }
}

File: IOS/OceanSentinels/Data/Local/Preferences/PreferencesManager.swift
import Foundation
import Combine
import os

/// UserDefaults-based preferences manager.
/// Replaces Android DataStore PreferencesManager.
///
/// Uses @AppStorage-compatible keys for SwiftUI binding.
/// Published properties emit Combine events for reactive updates.
final class PreferencesManager: ObservableObject {
    
    static let shared = PreferencesManager()
    
    private let defaults: UserDefaults
    private let logger = Logger(subsystem: "com.oceansentinels.app", category: "Preferences")
    
    // MARK: - Keys
    
    private enum Keys {
        static let authToken = "auth_token"
        static let userData = "user_data"
        static let isLoggedIn = "is_logged_in"
        static let rememberMe = "remember_me"
        static let darkMode = "dark_mode"
        static let notificationsEnabled = "notifications_enabled"
        static let apnsToken = "apns_token"  // Replaces FCM token
        static let firstLaunch = "first_launch"
        static let demoPopupShown = "demo_popup_shown"
        static let offlineMode = "offline_mode"
    }
    
    // MARK: - Published Properties
    
    @Published var authToken: String? {
        didSet { defaults.set(authToken, forKey: Keys.authToken) }
    }
    
    @Published var isLoggedIn: Bool {
        didSet { defaults.set(isLoggedIn, forKey: Keys.isLoggedIn) }
    }
    
    @Published var rememberMe: Bool {
        didSet { defaults.set(rememberMe, forKey: Keys.rememberMe) }
    }
    
    @Published var isDarkMode: Bool {
        didSet { defaults.set(isDarkMode, forKey: Keys.darkMode) }
    }
    
    @Published var notificationsEnabled: Bool {
        didSet { defaults.set(notificationsEnabled, forKey: Keys.notificationsEnabled) }
    }
    
    @Published var isFirstLaunch: Bool {
        didSet { defaults.set(isFirstLaunch, forKey: Keys.firstLaunch) }
    }
    
    @Published var hasDemoPopupShown: Bool {
        didSet { defaults.set(hasDemoPopupShown, forKey: Keys.demoPopupShown) }
    }
    
    @Published var isOfflineMode: Bool {
        didSet { defaults.set(isOfflineMode, forKey: Keys.offlineMode) }
    }
    
    // MARK: - Init
    
    private init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        
        // Load persisted values
        self.authToken = defaults.string(forKey: Keys.authToken)
        self.isLoggedIn = defaults.bool(forKey: Keys.isLoggedIn)
        self.rememberMe = defaults.bool(forKey: Keys.rememberMe)
        self.isDarkMode = defaults.bool(forKey: Keys.darkMode)
        self.notificationsEnabled = defaults.object(forKey: Keys.notificationsEnabled) as? Bool ?? true
        self.isFirstLaunch = defaults.object(forKey: Keys.firstLaunch) as? Bool ?? true
        self.hasDemoPopupShown = defaults.bool(forKey: Keys.demoPopupShown)
        self.isOfflineMode = defaults.bool(forKey: Keys.offlineMode)
    }
    
    // MARK: - Auth Token
    
    func saveAuthToken(_ token: String) {
        authToken = token
        logger.debug("Auth token saved")
    }
    
    func clearAuthToken() {
        authToken = nil
        logger.debug("Auth token cleared")
    }
    
    // MARK: - User Data (Codable → JSON)
    
    var userData: User? {
        get {
            guard let data = defaults.data(forKey: Keys.userData) else { return nil }
            return try? JSONDecoder().decode(User.self, from: data)
        }
        set {
            if let user = newValue {
                let data = try? JSONEncoder().encode(user)
                defaults.set(data, forKey: Keys.userData)
            } else {
                defaults.removeObject(forKey: Keys.userData)
            }
            objectWillChange.send()
        }
    }
    
    func saveUserData(_ user: User) {
        userData = user
        logger.debug("User data saved: \(user.username)")
    }
    
    func clearUserData() {
        userData = nil
        logger.debug("User data cleared")
    }
    
    // MARK: - APNs Token (replaces FCM)
    
    var apnsToken: String? {
        get { defaults.string(forKey: Keys.apnsToken) }
        set { defaults.set(newValue, forKey: Keys.apnsToken) }
    }
    
    func saveAPNsToken(_ token: String) {
        apnsToken = token
        logger.debug("APNs token saved")
    }
    
    // MARK: - Session Management
    
    func setLoggedIn(_ loggedIn: Bool) {
        isLoggedIn = loggedIn
    }
    
    func setFirstLaunchComplete() {
        isFirstLaunch = false
    }
    
    func setDemoPopupShown() {
        hasDemoPopupShown = true
    }
    
    // MARK: - Clear
    
    func clearSession() {
        authToken = nil
        userData = nil
        isLoggedIn = false
        logger.info("Session cleared")
    }
    
    func clearAll() {
        let domain = Bundle.main.bundleIdentifier ?? "com.oceansentinels.app"
        defaults.removePersistentDomain(forName: domain)
        defaults.synchronize()
        
        // Reset published properties to defaults
        authToken = nil
        isLoggedIn = false
        rememberMe = false
        isDarkMode = false
        notificationsEnabled = true
        isFirstLaunch = true
        hasDemoPopupShown = false
        isOfflineMode = false
        
        logger.info("All preferences cleared")
    }
}
